### ISSUE TITLE
ClickHouse operator does not reliably restore storage after you delete a PV to remove a bad volume

### ISSUE THREAD

**Issue by hodgesrm** [Altinity/clickhouse-operator#1439]
The 0.22 operator added support for deleting a volume that has become corrupted. This does not seem to work fully in 0.23.5 and there's also a bug in the test for this behavior: It does not check that persistent storage is actually created. 

I followed the procedure in the tests to delete the PV, which is located here: https://github.com/Altinity/clickhouse-operator/blob/ed8f7f0086b0ee60241248be0a6cfc7a6b8d9b2a/tests/e2e/test_operator.py#L3622. 

We start with a cluster named demo2 with 2 replicas running in AWS EKS. There is a ReplicatedMergeTree table with some data in it. First, list the pods and get the name of the second replica. 
```
$ kubectl get pods
NAME                 READY   STATUS    RESTARTS   AGE
chi-demo2-s3-0-0-0   1/1     Running   0          7d13h
chi-demo2-s3-0-1-0   1/1     Running   0          47m
```
List the PVCs for the replica and locate the PV name for this pod. 
```
$ kubectl get pvc
NAME                                                   STATUS   VOLUME                                     
clickhouse-keeper-datadir-volume-clickhouse-keeper-0   Bound    pvc-4ce08c0e-96b6-402e-8de6-b0aa6f8020d7
storage-chi-demo2-s3-0-0-0                             Bound    pvc-d3407d79-350f-4cf2-821e-f67bfbc9f170
storage-chi-demo2-s3-0-1-0                             Bound    pvc-5be34bbf-08ec-48fc-824d-5842b8ec20fe
```

Delete the PV. This command may hang, so we put it in background. 
```
$ kubectl delete pv pvc-5be34bbf-08ec-48fc-824d-5842b8ec20fe --force &
```

Delete the finalizers on the PV. This will allow it to be deleted. 
```
$ kubectl patch pv pvc-5be34bbf-08ec-48fc-824d-5842b8ec20fe --type='json' --patch='[{"op":"remove","path":"/metadata/finalizers"}]'
persistentvolume/pvc-5be34bbf-08ec-48fc-824d-5842b8ec20fe patched
```

Delete the PVC. 
```
kubectl delete pvc storage-chi-demo2-s3-0-1-0
```

Finally, patch the ClickHouseInstallation resource to force reconciliation by adding a random taskID. This will recreate the storage and add in the table schema. 
```
 kubectl patch chi demo2 --type=merge -p '{"spec":{"taskID":"25"}}'
```

At the end I expect that the storage will be recreated and any tables that previously existed will be recreated and populated from replication. However, I observe the following differing outcomes: 

1. The table was recreated and replicated but the storage was not actually persistent. When I looked for the new PV it had not been created. It appears the ClickHouse pod was running on ephemeral storage. 
2. The storage was created but the schema was not populated. 

It looks as if there's a race condition of some kind here. The test code includes what amount to waits. At this point I can't reliability recreate the storage.

**Comment by hodgesrm** [comment]
I should add it does not seem to be necessary to delete the PVC as the test case does not do that. This was just my last try when I was flailing around a bit.

**Comment by alex-zaitsev** [comment]
Test is correct. However, 2 out of 3 test paths fail in EKS (it works in minikube used for CI/CD).

1. Delete STS and PVC one works with no issues.

2. Delete PVC only does not work -- PV is not deleted despite the fact that RetentionPolicy=Delete. PV is left in Released state but does not get deleted. This violates k8s contract actually.
```
pvc-55cd52a9-f891-47de-a31b-0c89f4bdd8a0   1Gi        RWO            Delete           Released   test-036-7f347a40-2954-11ef-ba78-acde48001122/default-chi-test-036-volume-re-provisioning-simple-0-0-0   gp3-encrypted                4m
```

3. Delete PV only does not work as well. Operator correctly detects this and re-creates PVC, but PVC is left in Pending state forever. I guess this is because PV is not fully deleted until it is attached to the node:

```
NAME                                                       STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS    AGE
default-chi-test-036-volume-re-provisioning-simple-0-0-0   Pending                                                                        gp3-encrypted   5m49s
```

```
LAST SEEN   TYPE      REASON               OBJECT                                                      MESSAGE
7m6s        Warning   VolumeFailedDelete   persistentvolume/pvc-572a8643-14c9-4e01-94fc-b61365f2eef2   persistentvolume pvc-572a8643-14c9-4e01-94fc-b61365f2eef2 is still attached to node ip-10-3-77-68.ec2.internal
```

This one can be fixed by deleting a pod to make sure the volume is detached.

So the are two working scenarios for volume re-creation (at least in EKS): 
1) Delete PVC and STS
2) Or delete PV and restart a Pod to make sure the PV is unbounded. 

In both cases operator re-creates everything properly.
